### PR TITLE
GFORMS-974 - Add new expression to check a value against a regular ex…

### PR DESCRIPTION
**Pull request by josef-vlach**
…pression

### CODE REVIEW

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    case IsFalse | IsTrue                 => List(Valid)
+    case Contains(collection, value)      => validateExprs(collection, value)
+    case In(value, _)                     => validateValueField(value)
+    case DateBefore(left, right)          => validateDateExprs(left, right)
+    case DateAfter(left, right)           => validateDateExprs(left, right)
+    case MatchRegex(value, _)             => validateValueField(value)
+  }
+}
+
+sealed trait BooleanExprWrapperType {
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:86]: may be call it BooleanExprForwardRefChecker for clarity ?

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+import uk.gov.hmrc.gform.core.{ Invalid, Valid, ValidationResult }
+import uk.gov.hmrc.gform.core.ValidationResult.BooleanToValidationResultSyntax
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+
+class BooleanExprValidator(
+  indexLookup: Map[FormComponentId, Int],
+  wrapperType: BooleanExprWrapperType
+)(
+  idx: Int
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:27]: probably worth renaming idx to pageIdx for clarity?

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    case DateFormCtxVar(FormCtx(formComponentId)) => List(formComponentId)
+    case _                                        => Nil
+  }
+
+  private def validateIdIdx(id: FormComponentId) =
+    indexLookup
+      .get(id)
+      .map { idIdx =>
+        wrapperType
+          .isForwardReference(idIdx, idx)
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:50]: idIdx => formComponentIdx?

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          .validationResult(s"id '${id.value}' named in $wrapperType is forward reference, which is not permitted")
+      }
+      .getOrElse(Invalid(s"id '${id.value}' named in $wrapperType expression does not exist in a form"))
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:53]: Minor: "id '${id.value}' named in $wrapperType expression does not exist in the current form"

File: test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala [hunk 5]
```diff
@@ -18,7 +18,7 @@ package uk.gov.hmrc.gform.core
```
Comment (girishkamat) [inline test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala:18]: could rename this class to FormTemplateValidatorSpec, as its unit tests for FormTemplateValidator

File: test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala [hunk 6]
```diff
@@ -555,6 +547,43 @@ class TemplateValidatorSpec extends Spec {
...
+    val formComponentsB = List(mkFormComponent("fieldB", Value))
+    val baseSectionB = mkSection("sectionB", formComponentsB)
+    val invalidRef = FormCtx(FormComponentId("a"))
+    val forwardReferenceError = Invalid("id 'fieldB' named in validIf is forward reference, which is not permitted")
+    val invalidReferenceError = Invalid("id 'a' named in validIf expression does not exist in a form")
+    val table =
+      Table(
+        ("booleanExpr", "expected"),
+        (Equals(forwardRef, constant), forwardReferenceError),
+        (Equals(invalidRef, constant), invalidReferenceError),
```
Comment (girishkamat) [inline test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala:562]: Would be good to add a scenario for validIf in validators?

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala (hunk 1 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:86]: Well, original idea was a type which remembers in what type was `BooleanExpr` wrapped in.

That `isForwardReference` method is convenient to have here is just an after effect. I can put that method elsewhere if  you wish?

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala (hunk 2 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:27]: sure

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala (hunk 3 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:50]: sure

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala (hunk 4 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:53]: sure

File: test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala (hunk 5 above)
Comment (josef-vlach) [reply test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala:18]: Out of the scope of this ticket.

File: test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala (hunk 6 above)
Comment (josef-vlach) [reply test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala]: Done.

File: test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala (hunk 5 above)
Comment (girishkamat) [reply test/uk/gov/hmrc/gform/core/TemplateValidatorSpec.scala:18]: No worries

File: app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala (hunk 1 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/formtemplate/BooleanExprValidator.scala:86]: Ok, I get it. Lets leave it as it is then

**Review by girishkamat — APPROVED** [review]

**Pull request merged**